sklearn/impute/_base.py

```diff
@@ -517,10 +517,15 @@ def transform(self, X):
             valid_statistics_indexes = np.flatnonzero(valid_mask)
 
             if invalid_mask.any():
-                missing = np.arange(X.shape[1])[invalid_mask]
+                invalid_features = np.arange(X.shape[1])[invalid_mask]
                 if self.verbose != "deprecated" and self.verbose:
+                    # use feature names warning if features are provided
+                    if hasattr(self, "feature_names_in_"):
+                        invalid_features = self.feature_names_in_[invalid_features]
                     warnings.warn(
-                        "Skipping features without observed values: %s" % missing
+                        "Skipping features without any observed values:"
+                        f" {invalid_features}. At least one non-missing value is needed"
+                        f" for imputation with strategy='{self.strategy}'."
```

Comment:
For the sake of explicitness, might it be worth specifying the warning category?
```suggestion
                        "Skipping features without any observed values:"
                        f" {invalid_features}. At least one non-missing value is needed"
                        f" for imputation with strategy='{self.strategy}'.",
                        category=UserWarning,
```

Reply:
`UserWarning` has always been the default warning category. I don't think this is needed.